project_W/model.py

```diff
@@ -81,19 +128,82 @@ def delete_user(
     return (f"Successfully deleted user with email {user.email}"), 200
 
 
-def update_user_password(
-    user: User, new_password: str
-) -> Tuple[str, int]:
-    user.set_password_unchecked(new_password)
-    db.session.commit()
-    logger.info(f" -> Updated password of user {user.email}")
-    return "Successfully updated user password", 200
+def is_valid_email(email: str) -> bool:
+    allowedDomains = flask.current_app.config["loginSecurity"]["allowedEmailDomains"]
+    pattern = r"^\S+@"
+    if allowedDomains == []: pattern += r"([a-z0-9\-]+\.)+[a-z0-9\-]+"
+    else: pattern += r"(" + "|".join(allowedDomains) + r")"
+    pattern += r"$"
+    return re.match(pattern, email) is not None
+
+def is_valid_password(password: str) -> bool:
+    return re.match(r"^(?=.*[A-Z])(?=.*[a-z])(?=.*[0-9])(?=.*[^a-zA-Z0-9]).{12,}$", password) is not None
```

Comment:
Special characters are supposed to be matched by the wildcard `[^a-zA-Z0-9]`. Strictly speaking it matches everything that is not a letter or number. `a1.aaaaaaaaA` gets matched, I just tested it. Since positive Lookaheads are non consuming the ordering of the symbols should not be relevant at all if that's what you meant. I can warmly recommend [regex101](https://regex101.com/) for stuff like this ;)